backend/core/src/main/java/org/sonarsource/sonarlint/core/tracking/matching/MatchingSession.java

```diff
@@ -0,0 +1,118 @@
+    this.introductionDateProvider = introductionDateProvider;
+  }
+
+  public TrackedIssue matchWithKnownFinding(Path relativePath, RawIssue rawIssue) {
+    if (rawIssue.isSecurityHotspot()) {
+      return matchWithKnownSecurityHotspot(relativePath, rawIssue);
+    } else {
+      return matchWithKnownIssue(relativePath, rawIssue);
+    }
+  }
+
+  public TrackedIssue matchWithKnownSecurityHotspot(Path relativePath, RawIssue newSecurityHotspot) {
+    var trackedSecurityHotspot = matchWithKnownFinding(relativePath, remainingUnmatchedSecurityHotspotsPerFile, newSecurityHotspot);
+    securityHotspotsPerFile.computeIfAbsent(relativePath, f -> new ArrayList<>()).add(trackedSecurityHotspot);
+    relativePathsInvolved.add(relativePath);
+    return trackedSecurityHotspot;
+  }
+
+  private TrackedIssue matchWithKnownIssue(Path relativePath, RawIssue rawIssue) {
+    var trackedIssue = matchWithKnownFinding(relativePath, remainingUnmatchedIssuesPerFile, rawIssue);
+    issuesPerFile.computeIfAbsent(relativePath, f -> new ArrayList<>()).add(trackedIssue);
+    relativePathsInvolved.add(relativePath);
+    return trackedIssue;
+  }
+
+  private TrackedIssue matchWithKnownFinding(Path relativePath, Map<Path, List<KnownFinding>> remainingUnmatchedKnownFindingsPerFile, RawIssue newFinding) {
+    var remainingUnmatchedKnownFindings = remainingUnmatchedKnownFindingsPerFile.getOrDefault(relativePath, new ArrayList<>());
+    var localIssueMatcher = new IssueMatcher<>(new RawIssueFindingMatchingAttributeMapper(), new KnownIssueMatchingAttributesMapper());
+    var localMatchingResult = localIssueMatcher.match(List.of(newFinding), remainingUnmatchedKnownFindings);
+    return localMatchingResult.getMatchOpt(newFinding)
+      .map(knownFinding -> {
+        remainingUnmatchedKnownFindings.remove(knownFinding);
+        return updateKnownFindingWithRawIssueData(knownFinding, newFinding);
+      })
+      .orElseGet(() -> newlyKnownIssue(relativePath, newFinding));
+  }
+
+  public static TrackedIssue updateKnownFindingWithRawIssueData(KnownFinding knownIssue, RawIssue rawIssue) {
+    return new TrackedIssue(knownIssue.getId(), rawIssue.getMessage(), knownIssue.getIntroductionDate(),
+      false, rawIssue.getSeverity(), rawIssue.getRuleType(), rawIssue.getRuleKey(),
+      TextRangeUtils.getTextRangeWithHash(rawIssue.getTextRange(), rawIssue.getClientInputFile()),
+      TextRangeUtils.getLineWithHash(rawIssue.getTextRange(), rawIssue.getClientInputFile()), knownIssue.getServerKey(), rawIssue.getImpacts(), rawIssue.getFlows(),
+      rawIssue.getQuickFixes(), rawIssue.getVulnerabilityProbability(), null, rawIssue.getRuleDescriptionContextKey(),
+      rawIssue.getCleanCodeAttribute(), rawIssue.getFileUri());
+  }
+
+  private TrackedIssue newlyKnownIssue(Path relativePath, RawIssue rawFinding) {
+    var introductionDate = introductionDateProvider.determineIntroductionDate(relativePath, rawFinding.getLineNumbers());
+    if (introductionDate == null) {
```

Comment:
If we return Instant.now() instead of null from date provider, we can remove this check

Reply:
Good suggestion